src/participant/prompts/promptHistory.ts

```diff
@@ -0,0 +1,198 @@
+          // we only want to include the question asked, not the user's
+          // database and collection names in the history item.
+          break;
+        }
+      }
+    }
+
+    // eslint-disable-next-line new-cap
+    return vscode.LanguageModelChatMessage.Assistant(message);
+  }
+
+  private static _handleChatRequestTurn({
+    previousTurn,
+    currentTurn,
+    nextTurn,
+    connectionNames,
+    namespaceIsKnown,
+  }: {
+    previousTurn: vscode.ChatRequestTurn | vscode.ChatResponseTurn | undefined;
+    currentTurn: vscode.ChatRequestTurn;
+    nextTurn: vscode.ChatRequestTurn | vscode.ChatResponseTurn | undefined;
+    connectionNames: string[] | undefined;
+    namespaceIsKnown: boolean;
+  }): vscode.LanguageModelChatMessage | undefined {
+    if (previousTurn instanceof vscode.ChatResponseTurn) {
+      const responseIntent = (previousTurn.result as ChatResult).metadata
+        ?.intent;
+
+      if (responseIntent === 'askForNamespace' && namespaceIsKnown) {
+        // If the namespace is already known, skip responses to prompts asking for it.
+        return undefined;
+      }
+    }
+
+    if (
+      nextTurn instanceof vscode.ChatResponseTurn &&
+      nextTurn.result.errorDetails?.message === ParticipantErrorTypes.FILTERED
+    ) {
+      // If the response to this request led to a filtered error,
+      // we do not want to include it in the history
+      return undefined;
+    }
+
+    if (
+      currentTurn.prompt?.trim().length === 0 ||
+      connectionNames?.includes(currentTurn.prompt)
+    ) {
+      // When the message is empty or a connection name then we skip it.
+      // It's probably going to be the response to the connect step.
+      return undefined;
+    }
+
+    // eslint-disable-next-line new-cap
+    return vscode.LanguageModelChatMessage.User(currentTurn.prompt);
+  }
+
+  /** When passing the history to the model we only want contextual messages
+  to be passed. This function parses through the history and returns
+  the messages that are valuable to keep. */
+  static getFilteredHistory({
+    connectionNames,
+    history,
+    databaseName,
+    collectionName,
+  }: {
+    connectionNames?: string[]; // Used to scrape the connecting messages from the history.
+    history?: vscode.ChatContext['history'];
+    databaseName?: string;
+    collectionName?: string;
+  }): vscode.LanguageModelChatMessage[] {
+    const messages: vscode.LanguageModelChatMessage[] = [];
+
+    if (!history) {
+      return [];
+    }
+
+    const namespaceIsKnown =
+      databaseName !== undefined && collectionName !== undefined;
+    for (let i = 0; i < history.length; i++) {
+      const currentTurn = history[i];
+
+      let addedMessage: vscode.LanguageModelChatMessage | undefined;
+      if (currentTurn instanceof vscode.ChatRequestTurn) {
+        const previousTurn = i - 1 >= 0 ? history[i - 1] : undefined;
+        const nextTurn = i + 1 < history.length ? history[i + 1] : undefined;
```

Comment:
Just a question, do we use nextTurn to chek for filtered errors?

Reply:
could you clarify what you mean by that? we are checking the `nextTurn` here to filter out the requests which have led to errors (specifically just the ones filtered by responsible AI service at the moment)

Reply:
This is exactly what I was asking for :) Like to understand why we look ahead of the current turn. Wanted to make sure I understood that correctly.